### PR TITLE
Install middleware in ObjectAdapter (C++)

### PR DESCRIPTION
**Pull request by bernardnormier**
This is the C++ version of #2277 / #2278.

### CODE REVIEW

File: cpp/src/Ice/ObjectAdapterI.cpp [hunk 1]
```diff
@@ -901,6 +899,25 @@ Ice::ObjectAdapterI::ObjectAdapterI(
 void
 Ice::ObjectAdapterI::initialize(optional<RouterPrx> router)
 {
+    // shared_from_this() is available now and is called by `use`.
+
+    const LoggerPtr logger = _instance->initializationData().logger;
```
Comment (bernardnormier) [inline cpp/src/Ice/ObjectAdapterI.cpp:904]: Same order as in C#: first LoggerMiddleware, then ObserverMiddleware.

File: cpp/src/Ice/ObjectAdapterI.h [hunk 2]
```diff
@@ -151,10 +151,9 @@ namespace Ice
         IceInternal::ThreadPoolPtr _threadPool;
         const IceInternal::ServantManagerPtr _servantManager;
 
-        // There is no need to clear _dispatchPipeline during destroy because _dispatchPipeline does not hold onto this
-        // object adapter directly. It can hold onto a communicator that holds onto this object adapter, but the
-        // communicator will release this refcount when it is destroyed or when the object adapter is destroyed.
-        const ObjectPtr _dispatchPipeline; // must be declared after _servantManager
+        mutable ObjectPtr _dispatchPipeline;
```
Comment (bernardnormier) [inline cpp/src/Ice/ObjectAdapterI.h:154]: Seems nicer than using const-cast or than making dispatchPipeline non-const.

File: cpp/src/Ice/ObjectAdapterI.cpp [hunk 3]
```diff
@@ -538,10 +527,20 @@ Ice::ObjectAdapterI::findServantLocator(const string& prefix) const
...
 {
-    return _servantManager;
+    lock_guard lock(_mutex);
+    if (!_dispatchPipeline)
+    {
+        _dispatchPipeline = _servantManager;
+        while (!_middlewareFactoryStack.empty())
+        {
+            _dispatchPipeline = _middlewareFactoryStack.top()(_dispatchPipeline);
+            _middlewareFactoryStack.pop();
```
Comment (bernardnormier) [inline cpp/src/Ice/ObjectAdapterI.cpp:540]: We release the memory held by the std::function / lambda as soon as we no longer need it.

File: cpp/src/Ice/ObjectAdapterI.cpp [hunk 4]
```diff
@@ -538,10 +527,20 @@ Ice::ObjectAdapterI::findServantLocator(const string& prefix) const
...
+Ice::ObjectAdapterI::dispatchPipeline() const noexcept
 {
-    return _servantManager;
+    lock_guard lock(_mutex);
+    if (!_dispatchPipeline)
+    {
+        _dispatchPipeline = _servantManager;
+        while (!_middlewareFactoryStack.empty())
+        {
+            _dispatchPipeline = _middlewareFactoryStack.top()(_dispatchPipeline);
```
Comment (pepone) [inline cpp/src/Ice/ObjectAdapterI.cpp:539]: can we move `_dispatchPipeline` here?
```suggestion
            _dispatchPipeline = _middlewareFactoryStack.top()(std::move(_dispatchPipeline));
```

File: cpp/test/Ice/middleware/AllTests.cpp [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+using namespace std;
+using namespace Test;
+
+class Middleware final : public Ice::Object
+{
+public:
+    Middleware(Ice::ObjectPtr next, const string& name, list<string>& inLog, list<string>& outLog)
+        : _next(std::move(next)),
+          _name(name),
```
Comment (pepone) [inline cpp/test/Ice/middleware/AllTests.cpp:14]: Pass string by value and move it?

```suggestion
    Middleware(Ice::ObjectPtr next, string name, list<string>& inLog, list<string>& outLog)
        : _next(std::move(next)),
          _name(std::move(name)),
```

**Review by pepone — APPROVED** [review]

**Review by externl — APPROVED** [review]

**Pull request merged**